repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/samlpolicy/SamlAttributeMappingTest.groovy

```diff
@@ -0,0 +1,429 @@
+import org.opensaml.saml.saml2.core.Response as SamlResponse
+import org.rackspace.deproxy.Deproxy
+import org.rackspace.deproxy.Request
+import org.rackspace.deproxy.Response
+import org.spockframework.runtime.ConditionNotSatisfiedError
+import spock.lang.FailsWith
+
+import static features.filters.samlpolicy.util.SamlPayloads.*
+import static features.filters.samlpolicy.util.SamlUtilities.*
+import static framework.mocks.MockIdentityV2Service.createIdpJsonWithValues
+import static framework.mocks.MockIdentityV2Service.createMappingJsonWithValues
+import static javax.servlet.http.HttpServletResponse.SC_OK
+
+/**
+ * This functional test ensures the attribute mappings are being set correctly on the request and the response.
+ */
+class SamlAttributeMappingTest extends ReposeValveTest {
+    static samlUtilities = new SamlUtilities()
+    static xmlSlurper = new XmlSlurper()
+    static jsonSlurper = new JsonSlurper()
+    static MockIdentityV2Service fakeIdentityV2Service
+
+    def setupSpec() {
+        reposeLogSearch.cleanLog()
+
+        def params = properties.defaultTemplateParams
+        repose.configurationProvider.applyConfigs("common", params)
+        repose.configurationProvider.applyConfigs("features/filters/samlpolicy", params)
+
+        deproxy = new Deproxy()
+
+        fakeIdentityV2Service = new MockIdentityV2Service(properties.identityPort, properties.targetPort)
+        deproxy.addEndpoint(properties.targetPort, 'origin service', null, fakeIdentityV2Service.handler)
+        deproxy.addEndpoint(properties.identityPort, 'identity service', null, fakeIdentityV2Service.handler)
+
+        repose.start()
+        reposeLogSearch.awaitByString("Repose ready", 1, 30)
+
+        fakeIdentityV2Service.admin_token = UUID.randomUUID().toString()
+    }
+
+    def setup() {
+        fakeIdentityV2Service.resetHandlers()
+    }
+
+    @FailsWith(ConditionNotSatisfiedError)
+    def "the saml:response will be translated before being sent to the origin service"() {
+        given: "a mapping policy with a literal value and a path-based value in addition to the standard attributes"
+        def extAttribLiteral = "banana"
+        def extAttribLiteralValue = "phone"
+        def extAttribPath = "adventure"
+        def extAttribPathValue = "Mordor"
+        def mappingPolicy = createMappingJsonWithValues(
+                userExtAttribs: [(extAttribLiteral): extAttribLiteralValue, (extAttribPath): "{0}"],
+                remote: [[path: $/\/saml2p:Response\/saml2:Assertion\/saml2:Subject\/saml2:NameID\/@SPProvidedID/$]])
+
+        and: "a saml:response with a value at the path specified by the mapping policy"
+        def samlIssuer = generateUniqueIssuer()
+        def attribName = "Frodo Baggins"
+        def attribRole = "nova:cool_cat"
+        def attribDomain = "193083"
+        def attribEmail = "frodo.baggins@nowhere.abc"
+        def saml = samlResponse(issuer(samlIssuer) >> status() >> assertion(
```

Comment:
Yay function composition! Admittedly, it threw me for a look at first since you have methods creating `Closure`s which are then composed, and the methods are the things that take parameters. Also, I hadn't seen the syntax for this in Groovy yet.